nvtabular/workflow.py

```diff
@@ -734,83 +718,87 @@ def apply(
+            self.iterate_online(
+                dataset,
+                output_path=output_path,
+                shuffle=shuffle,
+                output_format=output_format,
+                out_files_per_proc=out_files_per_proc,
+            )
+
+    def iterate_online(
+        self,
+        dataset,
+        end_phase=None,
+        output_path=None,
+        shuffle=None,
+        output_format=None,
+        out_files_per_proc=None,
+        apply_ops=True,
+    ):
+        """ Iterate through dataset and (optionally) apply/shuffle/write.
+        """
+        # Check if we have a (supported) writer
+        output_path = output_path or "./"
+        output_path = str(output_path)
+        writer = nvt_io.writer_factory(output_format, output_path, out_files_per_proc, shuffle)
+
+        # Iterate through dataset, apply ops, and write out processed data
+        if apply_ops:
             for gdf in dataset.to_iter():
-                self.apply_ops(
-                    gdf,
-                    output_path=output_path,
-                    shuffler=shuffler,
-                    num_out_files=out_files_per_proc,
-                    huge_ctr=huge_ctr,
-                )
-            if shuffler:
-                shuffler.close()
-            if huge_ctr:
-                huge_ctr.close()
+                self.apply_ops(gdf, output_path=output_path, writer=writer, shuffle=shuffle)
 
-    def reorder_tasks(self, end):
-        if end != 2:
-            # Opt only works for two phases (for now)
-            return
-        stat_tasks = []
-        trans_tasks = []
-        for idx, _ in enumerate(self.phases[:end]):
-            for task in self.phases[idx]:
-                deps = task[2]
-                if isinstance(task[0], StatOperator):
-                    if deps == ["base"]:
-                        stat_tasks.append(task)
-                    else:
-                        # This statistics depends on a transform
-                        # (Opt wont work)
-                        return
-                elif isinstance(task[0], TransformOperator):
-                    trans_tasks.append(task)
+        # Close writer and write general/specialized metadata
+        if writer:
+            general_md, special_md = writer.close()
 
-        self.phases[0] = stat_tasks
-        self.phases[1] = trans_tasks
+            # Note that we "could" have the special and general metadata
+            # written during `writer.close()` (just above) for the single-GPU case.
+            # Instead, the metadata logic is separated from the `Writer` object to
+            # simplify multi-GPU integration. When using Dask, we cannot assume
+            # that the "shared" metadata files can/will be written by the same
+            # process that writes the data.
+            type(writer).write_special_metadata(special_md, writer.fs, output_path)
+            type(writer).write_general_metadata(general_md, writer.fs, output_path)
```

Comment:
Thanks for adding the comment!

fwiw, you can call classmethods on instances of that class so the ```type(writer)``` isn't strictly necessary (though obviously still works).

Reply:
Ah right - Forgot about that.